### PR TITLE
JDK-8280503: Use allStatic.hpp instead of allocation.hpp where possible

### PR DESCRIPTION
**Pull request by tstuefe**
JDK-8249944 moved AllStatic to its own header. We should use that one instead of allocation.hpp where possible to reduce header dependencies.

This patch:
- replaces includes of allocation.hpp with allstatic.hpp where appropiate
- fixes up resulting errors since this changes uncovers missing dependencies. Mainly, missing includes of debug.hpp, of globalDefinitions.hpp, and missing outputStream definitions.

Changes are trivial but onerous. Done partly with a script, partly manually.

Test:
- Checked the build with gtests on Linux x86, x64, minimal, zero, aarch64, for both fastdebug and release. All builds of course without PCH.
- GHAs

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8280503](https://bugs.openjdk.java.net/browse/JDK-8280503): Use allStatic.hpp instead of allocation.hpp where possible

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to 37fd8fdbb53972f0717de03212cd5ea47e0c8270
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7188/head:pull/7188` \
`$ git checkout pull/7188`

Update a local copy of the PR: \
`$ git checkout pull/7188` \
`$ git pull https://git.openjdk.java.net/jdk pull/7188/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7188`

View PR using the GUI difftool: \
`$ git pr show -t 7188`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7188.diff">https://git.openjdk.java.net/jdk/pull/7188.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/cds/archiveUtils.cpp [hunk 1]
```diff
@@ -39,7 +39,9 @@
 #include "oops/compressedOops.inline.hpp"
 #include "runtime/arguments.hpp"
 #include "utilities/bitMap.inline.hpp"
+#include "utilities/debug.hpp"
 #include "utilities/formatBuffer.hpp"
+#include "utilities/globalDefinitions.hpp"
```
Comment (dholmes-ora) [inline src/hotspot/share/cds/archiveUtils.cpp:44]: Seems unrelated to this issue.

**Review by dholmes-ora — APPROVED** [review]
Hi Thomas,

Seems okay - hard to validate (and I expect allocation.hpp to be included somewhere in most cases anyway).

Some of the changes seem to have nothing to do with:

-#include "memory/allocation.hpp"
+#include "memory/allStatic.hpp"

? Are these caused by transitive include changes?

Thanks,
David

**Review by iklam — COMMENTED** [review]
Looks good to me. I've validated with these builds locally on my machine: linux-x64-debug linux-aarch64-open-debug linux-arm32 linux-ppc64le-debug linux-s390x-debug linux-aarch64-debug linux-arm32-open-debug linux-aarch64-lic

I am running a mach5 job for tier1 + builds-tier5. That should cover most of the builds done by the Oracle CI. I'll post the results when they are ready.

**Comment by iklam** [comment]
BTW, I have some scripts for checking how often a header file is included. See https://github.com/iklam/tools/tree/main/headers

count_hotspot_headers.tcl shows that allocation.hpp was included by 1006 .o files before this fix, and 996 files afterwards, so not a whole lot of reduction. That's because we have over 300 headers that include allocatons.hpp :-)

**Comment by iklam** [comment]
> I am running a mach5 job for tier1 + builds-tier5. That should cover most of the builds done by the Oracle CI. I'll post the results when they are ready.

Unfortunately I am seeing failures on macos and windows:

macos:

src/hotspot/os/bsd/gc/z/zNUMA_bsd.cpp:25: src/hotspot/share/gc/z/zNUMA.hpp:39:10: error: unknown type name 'uint32_t'

windows:

src\hotspot\os\windows\threadLocalStorage_windows.cpp(34): error C3861: 'assert': identifier not found

File: src/hotspot/share/cds/archiveUtils.cpp (hunk 1 above)
Comment (tstuefe) [reply src/hotspot/share/cds/archiveUtils.cpp:44]: archiveUtils.cpp misses debug.hpp (since it uses assert) and globalDefinitions.hpp (since it uses types from there) but missed these includes. This had been hidden by one of its includes pulling allocation.hpp. Same goes for the other seemingly unrelated fixups: all missing includes hidden by including allocation.hpp

**Comment by tstuefe** [comment]
> Hi Thomas,
> 
> Seems okay - hard to validate (and I expect allocation.hpp to be included somewhere in most cases anyway).
> 

It shouldn't, that's the idea of JDK-8249944.

> Some of the changes seem to have nothing to do with:
> 
> -#include "memory/allocation.hpp" +#include "memory/allStatic.hpp"
> 
> ? Are these caused by transitive include changes?
> 

See my comment in code. All places that needed to be fixed.

> Thanks, David

Thanks for the review,

..Thomas

**Comment by tstuefe** [comment]
> BTW, I have some scripts for checking how often a header file is included. See https://github.com/iklam/tools/tree/main/headers
> 
> count_hotspot_headers.tcl shows that allocation.hpp was included by 1006 .o files before this fix, and 996 files afterwards, so not a whole lot of reduction. That's because we have over 300 headers that include allocatons.hpp :-)

Yes. allocation.hpp could be split up more. E.g. MEMFLAGS and the NMT categories really should live somewhere else, I saw some places where allocation.hpp was included only because of them. StackObj may also be a good candidate for moving to an own small header.

Does your tool tell you include chokepoints, maybe its just one central include pulling in allocation.hpp?

> Unfortunately I am seeing failures on macos and windows:
> 
> macos:
> 
> src/hotspot/os/bsd/gc/z/zNUMA_bsd.cpp:25: src/hotspot/share/gc/z/zNUMA.hpp:39:10: error: unknown type name 'uint32_t'
> 
> windows:
> 
> src\hotspot\os\windows\threadLocalStorage_windows.cpp(34): error C3861: 'assert': identifier not found

Strange, since the GHAs went through. Not good. The bugs are easy to fix though. Thanks for testing.

**Comment by iklam** [comment]
> > BTW, I have some scripts for checking how often a header file is included. See https://github.com/iklam/tools/tree/main/headers
> 
> Does your tool tell you include chokepoints, maybe its just one central include pulling in allocation.hpp?
> 

The whoincludes.tcl script can do that. Unfortunately it tells us that many popular header (such as ostream.hpp that was itself included 976 times) include allocations.hpp.

```
src/hotspot$ tclsh whoincludes.tcl allocation.hpp| head -20
scanning    997 allocation.hpp
   2 found    976 ostream.hpp
   3 found    960 exceptions.hpp
   4 found    938 atomic.hpp
   5 found    891 memRegion.hpp
   6 found    877 iterator.hpp
   7 found    871 arena.hpp
   8 found    864 mutex.hpp
   9 found    860 growableArray.hpp
  10 found    855 mutexLocker.hpp
  11 found    855 autoRestore.hpp
  12 found    848 padded.hpp
  13 found    841 linkedlist.hpp
  14 found    835 jfrAllocation.hpp
  15 found    832 resourceHash.hpp
  16 found    829 gcUtil.hpp
  17 found    825 threadHeapSampler.hpp
  18 found    825 thread.hpp
  19 found    825 filterQueue.hpp
  20 found    679 symbol.hpp
```

**Comment by tstuefe** [comment]
> > > BTW, I have some scripts for checking how often a header file is included. See https://github.com/iklam/tools/tree/main/headers
> > 
> > 
> > Does your tool tell you include chokepoints, maybe its just one central include pulling in allocation.hpp?
> 
> The whoincludes.tcl script can do that. Unfortunately it tells us that many popular header (such as ostream.hpp that was itself included 976 times) include allocations.hpp.
> 
> ```
> src/hotspot$ tclsh whoincludes.tcl allocation.hpp| head -20
> scanning    997 allocation.hpp
>    2 found    976 ostream.hpp
>    3 found    960 exceptions.hpp
>    4 found    938 atomic.hpp
>    5 found    891 memRegion.hpp
>    6 found    877 iterator.hpp
>    7 found    871 arena.hpp
>    8 found    864 mutex.hpp
>    9 found    860 growableArray.hpp
>   10 found    855 mutexLocker.hpp
>   11 found    855 autoRestore.hpp
>   12 found    848 padded.hpp
>   13 found    841 linkedlist.hpp
>   14 found    835 jfrAllocation.hpp
>   15 found    832 resourceHash.hpp
>   16 found    829 gcUtil.hpp
>   17 found    825 threadHeapSampler.hpp
>   18 found    825 thread.hpp
>   19 found    825 filterQueue.hpp
>   20 found    679 symbol.hpp
> ```

That's a useful script. Well, maybe we can break this up a bit more.

BTW, can you send me your configure line for your breaking builds? I'd like to reproduce them locally.

**Comment by tstuefe** [comment]
Hi Ioi,

I fixed Windows, but cannot test MacOS since I don't have the hardware. Could you please give this another try?

Found out that the reason we don't see failing builds in GHAs is that GHAs build with precompiled headers. We should change this if possible.

Thanks, Thomas

**Review by iklam — APPROVED** [review]
All builds in our CI passed.

**Comment by tstuefe** [comment]
> All builds in our CI passed.

Thanks a lot, Ioi!
